src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java

```diff
@@ -0,0 +1,522 @@
+                if (currentState.nodes().masterNode() == null) {
+                    return currentState;
+                }
+                RoutingAllocation.Result routingResult = allocationService.reroute(currentState);
+                if (!routingResult.changed()) {
+                    return currentState;
+                }
+                return ClusterState.builder(currentState).routingResult(routingResult).build();
+            }
+
+            @Override
+            public void onFailure(String source, Throwable t) {
+                logger.warn("failed to perform reroute for {}", t, source);
+            }
+        });
+    }
+
+    /**
+     * Fills the shard fetched data with new (data) nodes and a fresh NodeEntry, and removes from
+     * it nodes that are no longer part of the state.
+     */
+    private void fillShardCacheWithDataNodes(Map<String, NodeEntry<T>> shardCache, DiscoveryNodes nodes) {
+        // verify that all current data nodes are there
+        for (ObjectObjectCursor<String, DiscoveryNode> cursor : nodes.dataNodes()) {
+            DiscoveryNode node = cursor.value;
+            if (shardCache.containsKey(node.getId()) == false) {
+                shardCache.put(node.getId(), new NodeEntry<T>(node.getId()));
+            }
+        }
+        // remove nodes that are not longer part of the data nodes set
+        for (Iterator<String> it = shardCache.keySet().iterator(); it.hasNext(); ) {
+            String nodeId = it.next();
+            if (nodes.nodeExists(nodeId) == false) {
+                it.remove();
+            }
+        }
+    }
+
+    /**
+     * Finds all the nodes that need to be fetched. Those are nodes that have no
+     * data, and are not in fetch mode.
+     */
+    private Set<NodeEntry<T>> findNodesToFetch(Map<String, NodeEntry<T>> shardCache) {
+        Set<NodeEntry<T>> nodesToFetch = new HashSet<>();
+        for (NodeEntry<T> nodeEntry : shardCache.values()) {
+            if (nodeEntry.hasData() == false && nodeEntry.isFetching() == false) {
+                nodesToFetch.add(nodeEntry);
+            }
+        }
+        return nodesToFetch;
+    }
+
+    /**
+     * Are there any nodes that are fetching data?
+     */
+    private boolean hasAnyNodeFetching(Map<String, NodeEntry<T>> shardCache) {
+        for (NodeEntry<T> nodeEntry : shardCache.values()) {
+            if (nodeEntry.isFetching()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Async fetches data for the provided shard with the set of nodes that need to be fetched from.
+     */
+    protected abstract void asyncFetch(ShardId shardId, Set<NodeEntry<T>> nodesToFetch, MetaData metaData, long cacheGeneration);
+
+    /**
+     * An implementation of async shard fetch that list shard stores.
+     */
+    public static class Store extends AsyncShardFetch<TransportNodesListShardStoreMetaData.NodeStoreFilesMetaData> {
```

Comment:
I think `Store` and `Started` should not be internal classes, it makes the `AsyncShardFetch` file smaller and easier to digest if they are moved to their own files.